Improvements for ReadyToRun helpers in JIT value numbering.

erozenfeld opened the pull request:
The following ReadyToRun helpers are now treated similarly to their normal
counterparts in value numbering:
CORINFO_HELP_READYTORUN_NEW
CORINFO_HELP_READYTORUN_NEWARR_1
CORINFO_HELP_READYTORUN_ISINSTANCEOF
CORINFO_HELP_READYTORUN_CHKCAST
CORINFO_HELP_READYTORUN_STATIC_BASE

In particular, this allows CSE-ing calls to the last 3 of the above helpers when
possible. #3281 is an issue for CORINFO_HELP_READYTORUN_STATIC_BASE.

Compiler::fgValueNumberHelperCallFunc is refactored to reduce code
duplication.

Closes #3281.

erozenfeld commented:
@briansull @jkotas PTAL

erozenfeld commented:
/cc @dotnet/jit-contrib

jkotas commented:
`crossgen mscorlib` is failing...

erozenfeld commented:
I pushed a fix for x86 crossgen mscorlib failure. It was related to helpers for newarr. Odd that it wasn't failing on amd64.

briansull commented on src/jit/utils.cpp:
```diff
@@ -1194,6 +1194,9 @@ void HelperCallProperties::init()
 
         case CORINFO_HELP_NEW_CROSSCONTEXT:
         case CORINFO_HELP_NEWFAST:
+#ifdef FEATURE_READYTORUN_COMPILER
```
I don't think that we shodul be using #ifdef READYTORUN
we  should just remove these ifdefs as these values are unconditionally defined in corinfo.h
I also note that the current source base already contains a mixture of places where we omit these #ifdef's
and instead use  if (opts.IsReadyToRun())
(see Importer.cpp line 4456)

briansull replied:
@jkotas Please confirm

jkotas replied:
Yes, I think that the code would be more readable and maintainable without these `#ifdef FEATURE_READYTORUN` that are not really making a lot of difference.

briansull commented:
LGTM

erozenfeld commented:
I removed the #ifdefs in this change except for one (because gtEntryPoint is only available under the ifdef).

I also added several unit tests to the readytorun tests. The impact of this change:
TestStaticBaseCSE: The number of calls to CORINFO_HELP_READYTORUN_STATIC_BASE goes from 6 to 1;
TestIsInstanceOfCSE: The number of calls to CORINFO_HELP_READYTORUN_ISINSTANCEOF goes from 2 to 1;
TestCastClassCSE: The number of calls to CORINFO_HELP_READYTORUN_CHKCAST goes from 2 to 1;
TestRangeCheckElimination: array range check is eliminated

I also ran all CoreClr tests in ReadyToRun mode and verified that this change doesn't introduce any new failures.

I verified no diffs in normal SuperPMI.

@briansull @jkotas PTAL

jkotas commented:
:+1:

jkotas commented:
@dotnet-bot test this please

jkotas commented:
@dotnet-bot test this please

jkotas commented:
@dotnet-bot test this please
